OAK-9392: Improve resilience when primary becomes unavailable

mreutegg opened the pull request:
Introduces a new configuration that sets a socket timeout for operations on the clusterNodes collection.

Default is up for discussion. Initially I set it to 30 seconds.

mreutegg committed:
OAK-9392: Improve resilience when primary becomes unavailable

Work in progress

mreutegg committed:
OAK-9392: Improve resilience when primary becomes unavailable

Some additional tests

reschke reviewed (commented):
can we change the description so it also (potentially) applies to RDB?

mreutegg commented:
I consider this issue MongoDocumentStore specific, based on how the MongoDB Java driver interacts with the replica-set. I don't know whether a similar problem exists for RDB, but it would probably depend on the JDBC driver.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoSessionFactory.java:
```diff
@@ -37,22 +37,23 @@
 
     private final ClientSessionOptions options;
 
+    private final MongoClock clock;
```
I might not fully understand this, but MongoClock is somewhat confusing me. IIUC it is only used by MongoSessionFactory and serves as a replacement for the `clusterTime` and `operationTime` fields that were previously used (and now become obsolete). In that case, could the scope of `MongoClock` not actually be limited to be a private/inner class of MongoSessionFactory - or is the MongoClock indeed being used outside of this class?

mreutegg replied:
The only usage is test, which is the reason I pulled it out of MongoSessionFactory. Otherwise, it is somewhat of an implementation details.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDBClient.java:
```diff
@@ -0,0 +1,145 @@
+ * Simple struct that contains {@code MongoClient}, {@code MongoDatabase} and
+ * {@code MongoStatus}.
+ */
+final class MongoDBClient {
```
I was wondering if there would be an alternative name for `MongoDBClient`. Reason is that there is code that does `client.getClient()` which is a bit difficult to read, as you need to know that there are effectively 2 layers or 2 concepts of client (the MongoDBClient and the MongoClient).. IIUC this is like a `Holder` or an `Info` (but the latter is also a bit underappreciating the functionality of this class).. so, not sure what's a better name, but at least this name was initially not directly intuitive (as it collides with MongoClient)

mreutegg replied:
Agreed, will think of something better.

mreutegg replied:
Renamed `MongoDBClient` to `MongoDBConnection` in https://github.com/apache/jackrabbit-oak/pull/283/commits/10404cbdd52341ba517e57a0e671537fa8a8853d

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfiguration.java:
```diff
@@ -73,6 +73,12 @@
+     * Name of framework property to configure socket timeout for lease update
+     * operations on MongoDB.
+     */
+    private static final String FWK_PROP_LEASE_SO_TIMEOUT = "oak.mongo.leaseSocketTimeout";
```
To @reschke 's comment re naming/generalising : what I noticed is that the property key has `mongo` in it (to make it clear, it is only for mongo) - but this field does not. Perhaps it could be renamed to sth like `FWK_PROP_MONGO_LEASE_SO_TIMEOUT` (and the same would then also apply to other places/methods that use leaseSoTimeout and not specifically call it out as belonging to mongo

mreutegg replied:
Not sure I understand. Do you suggest turning the configuration `leaseSocketTimeout` into a generic one that also applies to RDB or rather making it fully MongoDB specific?

stefan-egli replied:
I meant making it fully MongoDB specific

reschke replied:
Ack; I misread the original ticket; primary/secondary config does not apply to RDB.

mreutegg replied:
Done with https://github.com/apache/jackrabbit-oak/pull/283/commits/da55aa3195fe63a249b19ebc86e3e7509cb92439

stefan-egli reviewed (commented):
Added some naming and design comments but otherwise lgtm

mreutegg committed:
OAK-9392: Improve resilience when primary becomes unavailable

Make ClusterNodeInfo.setClock() and resetClockToDefault() accessible to tests in other packages.

mreutegg committed:
OAK-9392: Improve resilience when primary becomes unavailable

Integration test using org.testcontainers and toxiproxy

mreutegg committed:
OAK-9392: Improve resilience when primary becomes unavailable

Clarify that the new lease socket timeout is MongoDB specific

mreutegg committed:
OAK-9392: Improve resilience when primary becomes unavailable

Rename MongoDBClient to MongoDBConnection

mreutegg committed:
OAK-9392: Improve resilience when primary becomes unavailable

Fix configuration entry for mongoLeaseSocketTimeout
